torch/_dynamo/variables/lists.py

```diff
@@ -662,3 +664,107 @@ def reconstruct(self, codegen):
 
 class TupleIteratorVariable(ListIteratorVariable):
     pass
+
+
+class SetVariable(VariableTracker):
+    def __init__(
+        self,
+        items: List[VariableTracker],
+        recursively_contains=None,
+        regen_guards=True,
+        **kwargs,
+    ):
+        underlying_items = kwargs.pop("_underlying_items", set())
+        super().__init__(recursively_contains=recursively_contains, **kwargs)
+        # Note - Set is still backed by a list, because we want set behavior over the contents,
+        assert isinstance(items, list)
+        assert all(isinstance(x, VariableTracker) for x in items)
+
+        # Sometimes, we know that we have passed in the guards from the items in the set
+        if regen_guards:
+            self.guards.update(VariableTracker.propagate(items)["guards"])
+
+        self.items = []
+        self._underlying_items = underlying_items
+        if underlying_items:
+            self.items = items
+        else:
+            self._add(items)
+
+    def as_proxy(self):
+        return [x.as_proxy() for x in self.items]
+
+    def python_type(self):
+        return set
+
+    def reconstruct(self, codegen):
+        codegen.create_load_python_module(builtins, True)
+        codegen.create_load_global("set", False)
+        codegen.foreach(self.items)
+        return [create_instruction("BUILD_SET", arg=len(self.items))]
+
+    # Note - this is only used for producing a set
+    def _to_example_values(self, vt):
+        from .base import VariableTracker
+        from .tensor import TensorVariable
+
+        assert isinstance(vt, VariableTracker)
+
+        if isinstance(vt, TensorVariable):
+            return vt.as_proxy().node.meta["example_value"]
+        if isinstance(vt, ConstantVariable):
+            return vt.value
+
+        unimplemented(f"Sets with {type(vt)} NYI")
+
+    def _add(self, item):
+        if isinstance(item, (list, set)):
+            for i in item:
+                self._add(i)
+            return self.items
+
+        ev = self._to_example_values(item)
+        if ev not in self._underlying_items:
```

Comment:
I'm trying to convince myself this will be correct for tensors.

My understanding is `set()` will call:
- `__hash__`: which is based on the `id()` of the tensor. Does `example_value` preserve aliasing relationships?
- `__eq__`: which reads tensors values, and throws an error with more than 1 value

Can you share your argument that using placeholder values will be correct?

Reply:
Nah, this is poor. 

I am going to use the source access I made for attrs to do this.

Reply:
@jansel see the above update to use `eval` to get the real value out for the same of finding uniques.

Reply:
Are we guarding on the aliasing relationships here?  What if the aliasing changes between calls?

If we properly guard on aliasing relationships, then we shouldn't need to use real values.